hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmBucketArgs.java

```diff
@@ -150,11 +151,20 @@ public DefaultReplicationConfig getDefaultReplicationConfig() {
     return defaultReplicationConfig;
   }
 
+  /**
+   * Returns true if defaultReplicationConfig has been set
+   * to a non default value.
+   */
+  public boolean hasDefaultReplicationConfig() {
```

Comment:
Do we actually need this method in this case? In #3975 the problem was a little different. The values there were longs, which in Java cannot be null and default to zero. Therefore it was impossible to tell if zero was passed or it was just the default value. The value here is an object and can be null, so maybe its enought to check it for non null?

In OMBucketSetProperty, we have:

```
      if (defaultReplicationConfig != null) {
        // Resetting the default replication config.
        bucketInfoBuilder.setDefaultReplicationConfig(defaultReplicationConfig);
      } else if (dbBucketInfo.getDefaultReplicationConfig() != null) {
        // Retaining existing default replication config
        bucketInfoBuilder.setDefaultReplicationConfig(
                  dbBucketInfo.getDefaultReplicationConfig());
      }
```
So we don't use this method there.

One thing that comes to mind - is it possible to remove a default rep config from a bucket, so it would then use the server default? Would we be able to?

Reply:
You're right @sodonnel we would not need `hasDefaultReplicationConfig()` in `OmBucketArgs`. `BucketArgs` object already has a `hasDefaultReplicationConfig()` that we can use to indicate if the replication config is set. 
If the replication config needs to be "reset"  i.e., removed from bucket args, would it make sense to reuse null default replication config to indicate both "to retain existing replication config" and "to reset to server default"? Would another non null indicator be useful make the distinction?

Reply:
If we use null for "don't overwrite" and also "clear" I am not sure how we can make the distinction.

To clear the replicationConfig, I think you will need:

1. A new parameter on OMBucketArgs - reset or clearReplicationConfig
2. Adjust RpcClient.setReplicationConfig, so that if a null ReplicationConfig is passed into it, it sets the new flag on OMBucketArgs.
3. Handle the new flag in the code where it saves the bucket args.
4. Change SetReplicationConfigHandler to either accept a "--reset" flag to indicate it need to clear the repConfig, or a new handler ResetReplicationConfigHandler to do the same, passing a null as ReplicationConfig to `bucket.setReplicationConfig(...)`

We can be fairly sure someone will come asking to clear the replicationConfig on a bucket some day, so it would be good to fix this. However I would do it in another Jira than this one, as I think this one is good now.